File: 46-permutations/js/main.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
let permute = function(nums) {
    let permutations = [];
    let current_permutation = [];
    let number_used_map = [];
    for (let index = 0; index < nums.length; index++) {
        number_used_map[index] = false;
    }

    recursive_permutation(nums, current_permutation, number_used_map, permutations);

    return permutations;
};

function recursive_permutation(nums, current_permutation, number_used_map, permutations) {
    if (current_permutation.length === nums.length) {
        // console.log("Current Permutation = " + current_permutation);
        // NOTE: Use concat function so that values are copied
        permutations.push(current_permutation.concat());
        // console.log(permutations);

        return;
    }

    for (let index = 0; index < nums.length; index++) {
        if (!number_used_map[index]) {
            number_used_map[index] = true;
            current_permutation.push(nums[index]);

            recursive_permutation(nums, current_permutation, number_used_map, permutations);

            current_permutation.pop();
            number_used_map[index] = false;
        }
    }
}


let nums = [1,2,3];
console.log("Permutations = ");
console.log(permute(nums));

File: 46-permutations/go/main.go
package main

import (
	"fmt"
	"slices"
)

func permute(nums []int) [][]int {
	var ans [][]int
	recursive_permute(0, nums, &ans)

	return ans
}

func swap(i int, j int, nums []int) {
	temp := nums[i]
	nums[i] = nums[j]
	nums[j] = temp
}

func recursive_permute(index int, nums []int, ans *[][]int) {
	if index >= len(nums) {
		permutation := slices.Clone(nums)
		*ans = append(*ans, permutation)

		return
	}

	for i := index; i < len(nums); i++ {
		swap(index, i, nums)
		recursive_permute(index + 1, nums, ans)
		swap(i, index, nums)
	}
}

func main() {
	nums := []int{1, 2, 3}
	fmt.Println("Nums are = ", nums)
	fmt.Println("Permutations are = ", permute(nums))
}
